source/extensions/filters/udp/dns_filter/dns_filter.h

```diff
@@ -0,0 +1,78 @@
+#pragma once
+
+#include "envoy/config/filter/udp/dns_filter/v2alpha/dns_filter.pb.h"
+#include "envoy/network/filter.h"
+
+#include "common/buffer/buffer_impl.h"
+#include "common/config/config_provider_impl.h"
+#include "common/network/utility.h"
+#include "common/runtime/runtime_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace UdpFilters {
+namespace DnsFilter {
+
+/**
+ * All Dns Filter stats. @see stats_macros.h
+ * Track the number of answered and un-answered queries for A and AAAA records
+ */
+#define ALL_DNS_FILTER_STATS(COUNTER)                                                              \
+  COUNTER(queries_a_record)                                                                        \
+  COUNTER(noanswers_a_record)                                                                      \
+  COUNTER(answers_a_record)                                                                        \
+  COUNTER(queries_aaaa_record)                                                                     \
+  COUNTER(noanswers_aaaa_record)                                                                   \
+  COUNTER(answers_aaaa_record)
+
+/**
+ * Struct definition for all Dns Filter stats. @see stats_macros.h
+ */
+struct DnsProxyFilterStats {
+  ALL_DNS_FILTER_STATS(GENERATE_COUNTER_STRUCT)
+};
+
+using DnsAddressList = std::vector<std::string>;
+using DnsVirtualDomainConfig = absl::flat_hash_map<std::string, DnsAddressList>;
+
+class DnsFilterEnvoyConfig {
+public:
+  DnsFilterEnvoyConfig(
+      Server::Configuration::ListenerFactoryContext& context,
+      const envoy::config::filter::udp::dns_filter::v2alpha::DnsFilterConfig& config);
+
+  DnsProxyFilterStats& stats() const { return stats_; }
+  DnsVirtualDomainConfig domains() const { return virtual_domains_; }
```

Comment:
do you mean to return a ref or a copy here?

Reply:
Let me try to re-work this.  The caller of the function stores the data in a reference, but I had problems actually returning a reference here.